### PR TITLE
Update to 5995 Element Controller Behavior

### PR DESCRIPTION
**Pull request by harmony7**
This is an additional fix for #5995.

My fix (PR #6026) "broke" cases such as #6035 of updating elements whose UIs were designed using the Forms API. It seems that these forms were the reason that the code before what I changed in #6026 included all of the acrobatics with the element's data dictionary and messing with `ValueProvider`.

I've made updates to `FormElementDriver` so that this case is handled. The code instantiates the involved form and iterates through their fields, copying the appropriate values out of the value provider and assigning them to the data dictionary. This I believe is a better approach than the previous way of simply copying Request.Form into the dictionary.

The other thing is, I had meant for this to be a fix for 1.9.x, since it doesn't break any interfaces and is what I would consider a "bug fix". If I'd made the previous PR against dev, then that was my mistake. I'm submitting both changesets again, as a pull request into 1.9.x.

### CODE REVIEW

**Comment by sfmskywalker** [comment]
This is looking good. Regarding 1.9.x, on the one hand you're right, but due to the nature of the changes, certain scenarios could break in unexpected ways (as @infofromca reported), so it is safer to merge it into dev. Also, despite the fact that your code is definitely an improvement, the current code in 1.9 is not causing issues as far as I know, so I think that is another reason to apply your changes to dev rather than 1.9.x.

Perhaps @sebastienros has a different opinion: 1.9.x or dev?

**Comment by harmony7** [comment]
@sfmskywalker i can see that. but please remember I started going down this path because of a bug: I had a problem writing a driver for my own element whose view model didn't have the same property names as the element. But sure let's wait for @sebastienros 's call

**Comment by sfmskywalker** [comment]
@harmony7 Ok. In any case, we think it's safest to merge it into dev in this case. If you could retarget your PR, that'd be awesome. Thanks!

File: src/Orchard.Web/Modules/Orchard.Layouts/Framework/Drivers/FormsElementDriver.cs [hunk 1]
```diff
@@ -32,8 +33,33 @@ public abstract class FormsElementDriver<TElement> : ElementDriver<TElement>, IF
...
+                        var name = ((dynamic) shape).Name as string;
+                        if (name != null) {
+                            var value = context.ValueProvider.GetValue(prefix + name);
+                            if (value != null) {
+                                context.Element.Data[name] = value.AttemptedValue;
+                            }
+                        }
+                    }));
+                }
+            }
```
Comment (jtkech) [inline src/Orchard.Web/Modules/Orchard.Layouts/Framework/Drivers/FormsElementDriver.cs]: @harmony7 after a quick try, indeed posted data are taken into account. Not sure but, as seen in my tests, when you re-edit the element, you don't retreive these data. Here, I needed to add this

            else // here when context.Updater == null
                context.ValueProvider = context.Element.Data.ToValueProvider(CultureInfo.InvariantCulture);

Note: maybe here it's better to use `_cultureAccessor.CurrentCulture` but I would have to change the `FormsElementDriver contructor` that is used in many places...

Best

**Comment by jtkech** [comment]
@harmony7 did you see my line note? As seen in my tests, when you re-edit the form element, you don't retreive its data. In `UpdateElementData()`, I needed to add this

        else // (when context.Updater == null)
            context.ValueProvider = context.Element.Data.ToValueProvider(CultureInfo.InvariantCulture);

Note: maybe here it's better to use `_cultureAccessor.CurrentCulture` but I would have to change the `FormsElementDriver contructor` that is used in many places...

Best

**Comment by harmony7** [comment]
Yeah I saw it, haven't had time to think about it yet. Thanks for the input.

**Comment by sfmskywalker** [comment]
I just worked on a fix that handles both updating the ElementData as per @harmony7 's suggestion as well as rebinding the form shape itself with the ElementData if not currently in a postback as per @jtkech 's suggestion. It's indeed best to use `ICultureAccessor`, but does break the base driver class, but that's easy to update, and since we're still on `dev` I think it's fine. To prevent breaking changes like this in the future I added a "services" interface that exposes `IFormManager` as well as `ICultureManager`, so that we can potentially add more properties to it without breaking the base driver class' signature.

**Pull request closed without merging**